db/fetcher/indexer.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/sourcenetwork/defradb/db/base"
+	"github.com/sourcenetwork/defradb/planner/mapper"
+)
+
+// IndexFetcher is a fetcher that fetches documents by index.
+// It fetches only the indexed field and the rest of the fields are fetched by the internal fetcher.
+type IndexFetcher struct {
+	docFetcher        Fetcher
+	col               *client.CollectionDescription
+	txn               datastore.Txn
+	indexFilter       *mapper.Filter
+	docFilter         *mapper.Filter
+	doc               *encodedDocument
+	mapping           *core.DocumentMapping
+	indexedField      client.FieldDescription
+	docFields         []client.FieldDescription
+	indexIter         indexIterator
+	indexDataStoreKey core.IndexDataStoreKey
+	execInfo          ExecInfo
+}
+
+var _ Fetcher = (*IndexFetcher)(nil)
+
+// NewIndexFetcher creates a new IndexFetcher.
+func NewIndexFetcher(
+	docFetcher Fetcher,
+	indexedFieldDesc client.FieldDescription,
+	indexFilter *mapper.Filter,
+) *IndexFetcher {
+	return &IndexFetcher{
+		docFetcher:   docFetcher,
+		indexedField: indexedFieldDesc,
+		indexFilter:  indexFilter,
+	}
+}
+
+func (f *IndexFetcher) Init(
+	ctx context.Context,
+	txn datastore.Txn,
+	col *client.CollectionDescription,
+	fields []client.FieldDescription,
+	filter *mapper.Filter,
+	docMapper *core.DocumentMapping,
+	reverse bool,
+	showDeleted bool,
+) error {
+	f.col = col
+	f.docFilter = filter
+	f.doc = &encodedDocument{}
+	f.mapping = docMapper
+	f.txn = txn
+
+	for _, index := range col.Indexes {
+		if index.Fields[0].Name == f.indexedField.Name {
+			f.indexDataStoreKey.IndexID = index.ID
+			break
+		}
+	}
+
+	f.indexDataStoreKey.CollectionID = f.col.ID
+
+	for i := range fields {
+		if fields[i].Name == f.indexedField.Name {
+			f.docFields = append(fields[:i], fields[i+1:]...)
+			break
+		}
+	}
+
+	iter, err := createIndexIterator(f.indexDataStoreKey, f.indexFilter, &f.execInfo)
+	if err != nil {
+		return err
+	}
+	f.indexIter = iter
+
+	if f.docFetcher != nil && len(f.docFields) > 0 {
+		err = f.docFetcher.Init(ctx, f.txn, f.col, f.docFields, f.docFilter, f.mapping, false, false)
+	}
+
+	return err
+}
+
+func (f *IndexFetcher) Start(ctx context.Context, spans core.Spans) error {
+	err := f.indexIter.Init(ctx, f.txn.Datastore())
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *IndexFetcher) FetchNext(ctx context.Context) (EncodedDocument, ExecInfo, error) {
+	totalExecInfo := f.execInfo
+	defer func() { f.execInfo.Add(totalExecInfo) }()
+	f.execInfo.Reset()
+	for {
+		f.doc.Reset()
+
+		indexKey, hasValue, err := f.indexIter.Next()
+		if err != nil {
+			return nil, ExecInfo{}, err
+		}
+
+		if !hasValue {
+			return nil, f.execInfo, nil
+		}
+
+		property := &encProperty{
+			Desc: f.indexedField,
+			Raw:  indexKey.FieldValues[0],
+		}
+
+		f.doc.key = indexKey.FieldValues[1]
+		f.doc.properties[f.indexedField] = property
+		f.execInfo.FieldsFetched++
+
+		if f.docFetcher != nil && len(f.docFields) > 0 {
```

Comment:
No, I didn't run any performance checks because it's not the indexer's job to do it.
The indexer's job is to fetch the indexed field and determine the dockey it belongs to. The rest is supposed to be done by a dedicated `DocumentFetcher` that does exactly that.

Moreover, it does some other stuff that is necessary, like running filter on the rest of the fields.

Reply:
> The indexer's job is to fetch the indexed field and determine the dockey it belongs to

That is not what it does though?  The indexer yields full documents.  Internally it is implemented so that it calls a private fetcher instance to get the other stuff non-indexed fields, but it does not need to be implemented that way.

Reply:
> That is not what it does though? 

I do not quite follow. Yes, int the end the indexers does the whole job of fetching the full requested document, but fetching of non-indexed fields is delegated.

> but it does not need to be implemented that way.

Of course it doesn't have to be this way. There are many ways of implementing this.
Do you have a specific concern about this solution or a suggestion?

Reply:
> Do you have a specific concern about this solution or a suggestion?

Yes, I was curious as to whether you had looked into the performance difference of using the inner fetcher vs just using txn.Get for each field, and considered whether such an implementation would be easier to read.

Reply:
sorry Andy, I perceived your comment originally a bit sceptical. But now I gave it another thought and it seems to be an interesting option to consider. 
I think though that it should be part of another task as this one is already large enough.
I've created a task to look into it a bit later https://github.com/sourcenetwork/defradb/issues/1936

Reply:
Commented on the new issue mentioned, not will add here.

I don't think a separate `Get()` method should be used. There's a number of things the fetcher can, should, and might do both now and in the future. I'd prefer not to implement the same or similar functionality in various places.

What is the concern for using the inner fetcher? Over fetching?

Reply:
> What is the concern for using the inner fetcher?

No concern, but `txn.Get` would almost certainly be much simpler, and I was curious as to what the performance diff would be like (I'd expect slight gains for small numbers of returned docs, with increasing losses as the number of returned docs increases).

Reply:
the whole idea of secondary indexes is about performance. If there a way to gain some more performance boost I think it should consider it.
But agree that adding a similar logic, or creating a new class is the most optimal way. But maybe we could add another configuration to the `DocumentFetcher` so that it fetches the value directly instead of seeking the next field every time. For some queries on a large data set it might prove quite beneficial I believe.

Reply:
If my memory is correct, Get is only faster for a very low number of calls, compared to Iterate.

Secondary indexes are indeed about performance, and from the benches it's doing a good job of it. 

It seems needlessly complex/redundant to duplicate the fetcher work, especially as they're are already several items in the works to improve the efficiency and perf of the fetcher. Additionally, as we add more CRDT types, there may be a need to add some more custom logic to the document fetching process, which would again need to be duplicated.

The cost of the fetcher can also be amortized by fetching with multiple spans (each span as a  separate doc key).

So at the moment I don't see the immediate need/benefit.